[1단계 - 자동차 경주 구현] 주드(홍성민) 미션 제출합니다.

kevstevie opened the pull request:
이번 미션은 대부분의 기능이 입력과 출력 로직이라 테스트 코드를 작성하는데 애매한 부분이 있어서 테스트 코드를 car 객체와 validator 사용시에만 작성했습니다.
만약 tdd를 진행하게 된다면 이런 부분도 전부 tdd를 하는지 궁금합니다.

kevstevie committed:
doc: README.md 기능 요구사항 작성

kevstevie committed:
feat: Car 클래스 구현 및 테스트코드 작성

kevstevie committed:
feat: 출력 메시지 상수 추가

kevstevie committed:
feat: 차 이름 입력 기능 구현

kevstevie committed:
feat: Junit study String 요구사항 1

kevstevie committed:
feat: Junit study String 요구사항 2

kevstevie committed:
feat: Junit study String 요구사항 3

kevstevie committed:
feat: Junit study set 요구사항 1

kevstevie committed:
feat: Junit study set 요구사항 2

kevstevie committed:
feat: Junit study set 요구사항 3

kevstevie committed:
refactor: Car 클래스의 move 메서드에 인수 추가

move 메서드에 int형 인수 random 추가
자동차의 초기 위치를 상수로 변경

kevstevie committed:
feat: 차 이름 입력 validation

kevstevie committed:
refactor: validateCarNames 접근 제어자 수정

kevstevie committed:
feat: 게임 초기화 로직 작성

kevstevie committed:
feat: 게임 시간 입력 유효성 검증 메서드 추가

kevstevie committed:
feat: 게임 시간 입력 메서드 추가

kevstevie committed:
feat: 랜덤한 값으로 차 위치를 업데이트한다

kevstevie committed:
feat: 차 위치 출력 메서드

kevstevie committed:
feat: 차 이동 및 중간 결과 출력 메서드

kevstevie committed:
refactor: OuputView 에서 OutputConstant 로 용도에 맞게 수정

kevstevie committed:
refactor: 역할에 맞게 메서드 수정

kevstevie committed:
feat: 우승자 출력 메서드

kevstevie committed:
refactor: 결과 출력에 필요한 name만 담도록 제네릭 수정

kevstevie committed:
feat: 결과 출력 메서드

kevstevie committed:
feat: 컨트롤러 게임 진행 로직

kevstevie committed:
feat: 차 이름 길이 에러메시지

kevstevie committed:
feat: 미입력 에러메시지

kevstevie committed:
feat: 숫자 미입력 에러 메시지

kevstevie committed:
feat: 양수 미입력 에러 메시지

kevstevie committed:
fix: 인터페이스 내 불필요한 접근제어자 제거

kevstevie committed:
feat: Validator Test 추가

KIMSIYOUNG commented on README.md:
```diff
@@ -5,3 +5,29 @@
 
 - [온라인 코드 리뷰 과정](https://github.com/woowacourse/woowacourse-docs/blob/master/maincourse/README.md)
+
+## 입력 기능
```
요구사항을 잘 정리해주셨네요. 
구현이 완료된 항목은 체크해주시면 좋을 것 같아요.
코드보다 문서는 더욱 Outdated되기 좋은 것 같아요. 
앞으로 쭉 잘 챙겨주시면 좋을 것 같습니다.

KIMSIYOUNG commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,29 @@
+        this.position = INIT_POSITION;
+    }
+
+    public void move(int random) {
```
자동차 객체 입장에서 고민해본다면 입력받는 값이 `random`하다는 것을 알 필요가 있을까요?

KIMSIYOUNG commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,66 @@
+
+import java.util.*;
+
+public class RacingGame {
```
해당 객체는 어떠한 역할이라고 생각하시나요?

해당 객체의 역할이 많은 것 같아요. 조금 더 분리해보면 좋을 것 같습니다.

KIMSIYOUNG commented on src/main/java/racingcar/util/Validator.java:
```diff
@@ -0,0 +1,57 @@
+
+import java.util.Arrays;
+
+public class Validator {
```
자동차 객체가 자신이 가져야하는 규칙을 직접 판단할 수 있지 않을까요? 앞으로 추가되는 모든 밸리데이션은 해당 객체에서 수행하게 되나요?

KIMSIYOUNG commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,26 @@
+
+public class InputView {
+
+
```
전체적으로 컨벤션은 한번씩 확인해주세요.

KIMSIYOUNG commented on src/main/java/racingcar/view/OutputConstant.java:
```diff
@@ -0,0 +1,8 @@
+
+public interface OutputConstant {
+
+    String INPUT_CAR_NAME = "경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).";
```
상수를 이렇게 분리하는게 어떤 이점이 있나요?

KIMSIYOUNG commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,23 @@
+
+    public void printWinners(List<String> winners) {
+        int beginIndex = 1;
+        int endIndex = winners.toString().length() - 1;
```
처음보는 사람이 이 로직을 이해할 수 있을까요?

KIMSIYOUNG commented on src/test/java/racingcar/CarTest.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    @Test
+    @DisplayName("4 이상의 랜덤 숫자를 받았을 때, 자동차 포지션이 1 증가하는지 검증한다")
```
4미만의 수를 받았을 때 전진하지 않는다도 요구사항인 것 같아요.

KIMSIYOUNG commented on src/test/java/racingcar/util/ValidatorTest.java:
```diff
@@ -0,0 +1,49 @@
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class ValidatorTest {
```
테스트와 TDD를 동일한 의미로 생각하시는 것 같아요. 
일단 이 둘이 무엇이 다른지를 파악하는게 필요할 것 같습니다.
추가로 지금 고민하시는 부분은 `무엇을 테스트할 것인지`인 것 같아요.
 
제 생각을 전달드리기보단, 말씀하신대로 이걸 테스트해야하는가? 한다면 어떻게 할 수 있는가? 를 고민해보시고 저한테 얘기해주시면 좋을 것 같습니다.

kevstevie replied:
저 스스로 생각해봤을때 유닛테스트는 도메인에 대해서 이루어지는게 테스트 용이성이나 테스트 효율에 있어서 맞다고 생각했습니다! 입력부분도 결국 도메인에 적용되기 때문에 이 방식으로 테스트 코드를
작성하면 좋을 것 같다고 생각해서 그렇게 수정했습니다

KIMSIYOUNG replied:
잘 고민해보신 것 같습니다. 
말씀하신대로 입력 그 자체도 테스트할 수 있는 방법은 있습니다.
다만 이걸 테스트해야하는지에 대해서 고민해보고 본인만의 기준으로 테스트 대상을 판단하는 것이 중요할 것 같아요.

추가로 입력부분도 결국 도메인에 적용된다는 의미는 어떤 의미일까요?

kevstevie replied:
입력받은 값이 도메인의 생성자에 적용되는 로직이기 때문에 도메인 검증을 통해 확인할 수 있다는 의미였습니다.

KIMSIYOUNG commented on src/test/java/racingcar/util/ValidatorTest.java:
```diff
@@ -0,0 +1,49 @@
+
+
+    @Test
+    @DisplayName("시도 횟수 공백 입력")
```
@Parameterized 키워드도 학습해보시면 좋을 것 같습니다.

KIMSIYOUNG reviewed (changes requested):
안녕하세요.
엄청 빠르게 잘 구현해주셨네요.
몇가지 코멘트 남겼는데 확인 부탁드립니다.

kevstevie committed:
doc: 요구 사항 1차 정리

kevstevie committed:
chore: 컨벤션에 맞게 정리

kevstevie committed:
refactor: car.move 파라미터 random 에서 number 로 변경

kevstevie committed:
feat: carTest @ParameterizedTest 적용해서 구현

kevstevie committed:
refactor: 우승자 산출시 최대값 찾는 로직 stream으로 수정

kevstevie committed:
refactor: String.join으로 승자 출력 가시성 개선

kevstevie committed:
refactor: 안티패턴인 상수 인터페이스 클래스로 수정

각 클래스에서 static import해서 사용하도록 함
sc 약어 변수 수정

kevstevie committed:
refactor: Car 생성자에서 이름 유효성 검증

kevstevie committed:
refactor: Car 생성자에서 검증하기 위해 검증 로직 수정

kevstevie committed:
refactor: RacingGame 도메인 객체와 gameController 클래스로 분리

kevstevie committed:
chore: 클래스 대문자 수정

kevstevie committed:
feat: 게임 마무리 메서드 추가

kevstevie committed:
feat: 승자산출 로직 추가

kevstevie committed:
doc: 변경 사항 반영

kevstevie committed:
feat: 게임 진행 로직 추가

kevstevie committed:
refactor: 게임 시간 입력 검증을 도메인과 인풋에서 나눠서 하도록 수정

kevstevie committed:
refactor: car 생성자에 trim 적용

kevstevie committed:
refactor: validationTest에 @Parameterized 적용

kevstevie committed:
chore: 매직넘버 상수 분리, 임포트 제거

kevstevie committed:
chore: 매직넘버 상수 분리

kevstevie commented:
리뷰 감사합니다!
테스트에 @ParameterizedTest 사용해서 여러개의 테스트 메서드를 작성하는 것보다 쉽게 여러 케이스 적용했습니다.

상수를 분리한 인터페이스는 찾아보니 안티패턴이라고 규정되어 있어서 final class 로 static import해서 사용했습니다.
상수를 모아놓은 클래스를 이용한 것은 추후 반복되는 메시지 관리에 이점이 있을거라 생각해서 게임 진행 메시지, 에러 메시지는 따로 클래스를 만들어 관리하도록 했습니다.

Validator 클래스도 마찬가지로 객체가 스스로 검증하는 메서드를 가지는 것보다 따로 관리해주는 객체가 있고 그것을 가져다 사용하는 것이 낫다고 생각해서 도메인 객체의 생성자에서 Validator를 사용하도록 수정했습니다.

맨 처음 RacingGame 객체를 설계할땐 Controller로서 작동하도록 만들었는데 cars와 winners를 가지는 도메인과 이를 컨트롤하는 두가지 객체로 역할을 분리했습니다.

kevstevie committed:
feat: 차를 하나만 입력한 케이스, 차이름이 중복인 케이스 예외처리

kevstevie committed:
refactor: 메인 애플리케이션 리네임

kevstevie committed:
refactor: winners를 인스턴스로 갖지 않고 승자 산출 로직으로 반환

KIMSIYOUNG commented on src/main/java/racingcar/util/Validator.java:
```diff
@@ -0,0 +1,76 @@
+    private Validator() {
+    }
+
+    public static void validateCarNames(String carName) {
```
말씀하신 방법도 코드를 유지보수하기 좋게하는 방향중 하나인 것 같아요.

다만 객체지향을 학습하고 있다면 객체가 가져야할 역할을 기준으로 분리해보면 어떨까요?
추가로 지금은 프로그램이 작지만, 프로그램이 커진다면 `Validator`는 어떻게 변화하게 될지도 고민해보면 좋을 것 같습니다.
Car와 CarValidator 두가지를 함께 가져간다면 `Car`를 이해하기 위해서 `CarValidator`가 함께 있어야하는데 충분히 복잡하지 않다면 오히려 관심사가 분리되어 파악하기 어렵지 않을까요?

kevstevie replied:
Validator와 ErrorMessage와 마찬가지로 결국 프로그램이 방대해진다면 OutputConstant의 상수들도 출력을 담당하는 클래스에서 가지고 있는게 맞겠군요

KIMSIYOUNG commented on src/main/java/racingcar/view/ErrorMessage.java:
```diff
@@ -0,0 +1,12 @@
+package racingcar.view;
+
+public final class ErrorMessage {
+    public static final String ERROR_PREFIX = "[ERROR] ";
```
ditto

KIMSIYOUNG commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,24 @@
+
+public class InputView {
+
+    private final Scanner scanner = new Scanner(System.in);
```
상수로 선언해도 괜찮을 것 같아요.

KIMSIYOUNG commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,24 @@
+    public int inputGameTime() {
+        System.out.println(INPUT_TRY_COUNT);
+        String input = scanner.nextLine();
+        int inputNumber = Validator.validateStringToInt(input);
```
Validator는 모두가 사용할 수 있는 것인가요? 뷰와 도메인 모두 의존하고 있는 것 같아서요.

KIMSIYOUNG commented on src/main/java/racingcar/view/OutputConstant.java:
```diff
@@ -0,0 +1,9 @@
+package racingcar.view;
+
+public final class OutputConstant {
```
ditto

KIMSIYOUNG commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,49 @@
+    private final List<Car> cars;
+    private final int gameTime;
+
+    public RacingGame(int gameTime, List<Car> cars) {
```
RacingGame의 역할을 아직 이해하지 못했어요. 조금 더 구체적으로 설명해주실 수 있나요?

해당 클래스가 있어야 하는 이유는 무엇인가요?

KIMSIYOUNG commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,49 @@
+    private static final int RANGE_MAX = 9;
+    private static final int RANGE_MIN = 0;
+
+    private final List<Car> cars;
```
일급컬렉션도 학습해보면 좋을 것 같아요.

KIMSIYOUNG commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,49 @@
+        return (int) ((Math.random() * (RANGE_MAX - RANGE_MIN)) + RANGE_MIN);
+    }
+
+    public List<String> calculateWinners() {
```
Winner 라는 객체를 만들어보는건 어떨까요?

KIMSIYOUNG commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,49 @@
+        cars.forEach(car -> car.move(generateRandomNumber()));
+    }
+
+    private int generateRandomNumber() {
```
아직 해당 객체의 역할이 많은 것 같아요. 조금 더 분리해보면 좋을 것 같아요.

KIMSIYOUNG reviewed (changes requested):
안녕하세요.
리뷰 잘 반영해주셨네요.
몇가지 추가 코멘트 남겼는데 확인 부탁드립니다.

kevstevie committed:
doc: 변경사항2 정리

kevstevie committed:
refactor: scanner 상수화, gameTime 입력 메서드 String 반환

kevstevie committed:
refactor: Validator 삭제, Car 객체에서 메서드로 직접 유효성 검증

kevstevie committed:
doc: Validator 삭제, Car 객체에서 메서드로 직접 유효성 검증

kevstevie committed:
feat: Cars 일급컬렉션 및 유효성검증 추가

kevstevie committed:
refactor: initCars 로직 수정

kevstevie committed:
feat: Cars에서 랜덤으로 car 객체를 움직이게 하는 메서드

kevstevie committed:
feat: Winner와 Winners 객체 추가

kevstevie committed:
refactor: 승자를 unmodifiableList로 반환하게 수정

kevstevie committed:
refactor: RacingGame, Validator 삭제

일급컬렉션이 로직을 가지고 Controller가 gameTime을 가지면서
RacingGame의 역할이 모호해짐

각 객체에서 필요한 유효성 검증을 직접 하면서 Validator의 역할이 사라짐

kevstevie committed:
refactor: cars를 관리하는 역할 및 gameTime을 인스턴스로 가지도록 수정

kevstevie committed:
refactor: gameTime을 컨트롤러에서 유효성 검증하도록 수정

kevstevie committed:
refactor: Winners 객체 적용하여 종료 로직 수정

kevstevie committed:
feat: getter 추가

kevstevie committed:
refactor: 패키지 구조 수정

kevstevie committed:
refactor: 생성자 유효성 검증 추가, 에러메시지 추가

kevstevie committed:
refactor: ErrorMessage 클래스 삭제, 각 도메인에 가지게 됨

kevstevie committed:
chore: 컨벤션 수정

kevstevie committed:
refactor: 메서드명 수정

kevstevie committed:
refactor: 메서드명 수정

kevstevie committed:
doc: 문서 최신화

kevstevie commented:
validator의 의존성에 대해 고민하다 유효성 검증이 필요한 각 클래스로 나누는 것을 선택했습니다.
다형성을 활용해서 Validator 인터페이스를 만들어 각 클래스의 Validator를 새로 만드는 방법도 생각해봤는데 피드백과 상충하는 것 같습니다.
이 방법이 추후의 확장성이나 Validator가 필요한 객체에서 유효성 검증 역할을 덜어준다는 측면에서 좋아보이는데 좀 혼란스럽습니다..
추가) 자문자답처럼 되었는데 "원시타입을 랩핑한다"라는 토픽을 학습하는 과정에서 다른 객체에 의존하는 것보다 스스로 검증하는 방식이 더 간결하다고 느끼게 되었네요.

RacingGame 객체를 통해 car 객체들의 움직임을 관리하고 시도횟수를 저장하면서 전반적인 게임 관리의 역할을 맡기고 있었는데
시도 횟수를 컨트롤러로 넘기고 car 객체 관리를 cars 일급 컬렉션으로 넘겨주면서 역할이 사라져서 아예 삭제했습니다.
리팩터링하고보니 기존 RacingGame의 대부분의 역할을 cars 컬렉션이 가지게 됐는데 이게 맞는걸까요?..

OutputConstant는 Validatord와 달리 뷰에서만 의존되고 있어서 유지하고 있습니다.

그리고 기존의 List<String>으로 승자 출력 파라미터로 넘겨주는 로직보다 Winner 객체를 만들었을때 얻는 이점이 무엇인지 잘 와닿지 않습니다.
추가) 승자를 출력하는 뷰에서 car 객체에 대한 의존이 사라지고 오직 winner 객체를 출력하는데에만 관심사를 가질 수 있다! 라고 접근하게 되었는데 이 부분이 맞을까요?

kevstevie committed:
refactor: 상수 접근 제어자 일괄 수정

kevstevie committed:
feat: Cars 예외 테스트

kevstevie committed:
feat: GameTime 객체 도입으로 원시타입 포장

kevstevie committed:
refactor: winners.toString으로 결과 출력하게 수정

kevstevie commented:
추가 질문이 있습니다. Winners 일급 컬렉션에서 getWinners를 하고 그 winners에서 또 getWinner를하고 또 다시 getName을 해서 뷰에 파라미터를 넘기는 방식이 랩핑을 하는 의미가 퇴색된다고 생각해서 toString을 오버라이딩하는 방식을 적용했는데 이 또한 도메인에서 출력방식을 결정하기 때문에 적절하지 않다고 생각이 드는데 마땅한 방법이 잘 떠오르지 않네요.

kevstevie committed:
feat: 승자 산출 로직 검증

kevstevie committed:
feat: 시도횟수 정상 입력 테스트 케이스

kevstevie committed:
refactor: Random 객체 상수로 변경

kevstevie committed:
chore: DELIMITER 상수 선언

KIMSIYOUNG commented on src/main/java/racingcar/ApplicationController.java:
```diff
@@ -0,0 +1,14 @@
+package racingcar;
+
+public class ApplicationController {
```
ApplicationController 와 GameController 또한 분리될 이유가 없는 것 같아요.
하나의 메소드에서 전체 프로세스를 진행해도, 메소드만 잘 분리되어 있다면 충분히 이해하기 좋은 코드일 것 같아요.

KIMSIYOUNG commented on src/main/java/racingcar/GameController.java:
```diff
@@ -0,0 +1,60 @@
+        return new GameTime(gameTime);
+    }
+
+    private List<Car> initCars() {
```
현재는  List<Car>를 Cars로 넣어주고 있는데요. 자동차들의 이름만 받아서 Cars에서 이를 수행해주어도 자연스러울 것 같습니다.

KIMSIYOUNG commented on src/main/java/racingcar/GameController.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public void runGame() {
+        while (gameTime.isNotZero()) {
```
gameTime이라는 유의미한 객체를 만들어주셨다면, isNotZero()라는 수동적인 네이밍 보단, isProgress() 와 같이 스스로 표현할 수 있게 해주면 더 좋을 것 같네요.

KIMSIYOUNG commented on src/main/java/racingcar/GameController.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public void finishGame() {
+        Winners winners = new Winners(cars.calculateWinners());
```
Winners 객체가 존재한다면 Cars객체에서 위너를 계산하는건 자연스럽지 않은 것 같아요.

Winners 또한 자신이 담당하고 있는 역할을 수행할 수 있게 해주면 어떨까요? (자동차들을 받아서 승자를 계산하는 로직을 가지는 형태가 되면, 객체 자체의 의미를 가질 수 있겠네요)

KIMSIYOUNG commented on src/main/java/racingcar/domain/Winners.java:
```diff
@@ -0,0 +1,26 @@
+        return winners;
+    }
+
+    public String toString() {
```
현재와 같은 구현은 사실상 객체가 아니라, 데이터 덩어리라고 표현하는게 적절할 것 같습니다. 그렇기에 포장한 이유가 와닿으시지 않는 것 같아요.

KIMSIYOUNG commented on src/main/java/racingcar/domain/Winner.java:
```diff
@@ -0,0 +1,14 @@
+package racingcar.domain;
+
+public class Winner {
```
ditto

KIMSIYOUNG commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,59 @@
+    private static final int DEFAULT_MAX = 0;
+
+    private static final Random random = new Random();
+    private List<Car> cars;
```
final도 선언해주면 좋을 것 같아요.

KIMSIYOUNG commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,50 @@
+        this(name, INIT_POSITION);
+    }
+
+    public Car(String name, int position) {
```
해당 생성자는 테스트에서 위치를 자유롭게 제어하기 위해서 열려있네요. 
테스트를 위한 프로덕션 코드 변경은 지양하는게 좋을 것 같습니다.
오히려 테스트하기 어렵다면, 내가 코드를 다르게 짤 방법은 없을까? 고민할 수 있을 것 같아요.

현재 잘 구현해주신 것처럼, 랜덤한 값을 Cars에서도 파라미터로 받을 수 있다면 해당 생서자는 필요없을 것 같아요.
또는 전략패턴, 인터페이스 라는 키워드로 학습해보셔도 좋을 것 같습니다.

KIMSIYOUNG commented on src/main/java/racingcar/GameController.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class GameController {
```
말씀하신대로 RacingGame이라는 값이 애초에 필요했던 값이 아니라고 저는 생각합니다.
자동차들을 관리하는 역할을 담당하는, Cars에서 모두 처리할 수 있었고 그게 더 맞는 방법인데 RacingGame이라는 객체가 필요한가? 라는 의도로 말씀드렸어요.

말씀주신 말투에서는 의구심이 있으신 것 같은데 어떤 부분이 마음에 걸리시는걸까요?

KIMSIYOUNG commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,18 @@
+
+import static racingcar.view.OutputConstant.RESULT_POSTFIX;
+
+public class OutputView {
```
말씀하신대로 값이 포장될 수록, 뷰에서는 값을 많이 꺼내와야하는 형태가 됩니다. 
해당 부분이 문제가 된다고 느끼는 이유가 있으실까요?

KIMSIYOUNG reviewed (approved):
안녕하세요 주드! 
피드백 드린 내용을 너무 잘 반영해주셔서 깜짝 놀랐네요.
몇가지 추가 코멘트 드렸는데 확인 부탁드립니다.
1단계는 여기서 머지하고 2단계에서 피드백 내용과 함께 조금 더 얘기해보면 좋을 것 같습니다.

KIMSIYOUNG commented on src/main/java/racingcar/GameTime.java:
```diff
@@ -0,0 +1,45 @@
+package racingcar;
```
요 친구는 도메인이 아닌가요?🙄

KIMSIYOUNG commented on src/main/java/racingcar/GameTime.java:
```diff
@@ -0,0 +1,45 @@
+    private static final int MAX_GAME_TIME = 500;
+    private static final int RANGE_BOUNDARY = 0;
+
+    private int gameTime;
```
final 키워드도 학습해보면 좋을 것 같아요.
